File: Src/Mod/DTG/WashingtonBaltimore/RailVehicles/Diesel/P42DC/Amtrak/Locomotive/P42DC_EngineScript.lua
-- Engine script for the P42DC operated by Amtrak.
--
-- @include RollingStock/BrakeLight.lua
-- @include SafetySystems/AspectDisplay/Genesis.lua
-- @include SafetySystems/Alerter.lua
-- @include Flash.lua
-- @include Iterator.lua
-- @include Misc.lua
-- @include RailWorks.lua
-- @include Units.lua
local adu
local alerter
local blight
local ditchflasher

local lasthorntime_s = nil
local lastthrottletime_s = nil

local function setenginenumber()
  local number = tonumber(RailWorks.GetRVNumber()) or 0
  RailWorks.SetControlValue("LocoHundreds", 0, Misc.getdigit(number, 2))
  RailWorks.SetControlValue("LocoTens", 0, Misc.getdigit(number, 1))
  RailWorks.SetControlValue("LocoUnits", 0, Misc.getdigit(number, 0))
end

Initialise = Misc.wraperrors(function()
  setenginenumber()

  local onebeep_s = 0.25
  adu = GenesisAdu:new{
    isamtrak = true,
    alerttone_s = onebeep_s,
    getbrakesuppression = function()
      return RailWorks.GetControlValue("TrainBrakeControl", 0) >= 0.4
    end,
    getacknowledge = function()
      return RailWorks.GetControlValue("AWSReset", 0) > 0
    end
  }

  alerter = Alerter:new{
    getacknowledge = function()
      return RailWorks.GetControlValue("AWSReset", 0) > 0
    end
  }
  alerter:start()

  blight = BrakeLight:new{}

  local ditchflash_s = 0.65
  ditchflasher = Flash:new{off_s = ditchflash_s, on_s = ditchflash_s}

  RailWorks.BeginUpdate()
end)

local function readcontrols()
  if RailWorks.GetControlValue("Horn", 0) > 0 then
    lasthorntime_s = RailWorks.GetSimulationTime()
  end
end

local function writelocostate()
  local penalty = alerter:ispenalty() or adu:ispenalty()
  -- There's no virtual throttle, so just move the combined power handle.
  if penalty then RailWorks.SetControlValue("ThrottleAndBrake", 0, 0.5) end
  -- There's no virtual train brake, so just move the braking handle.
  if penalty then RailWorks.SetControlValue("TrainBrakeControl", 0, 0.85) end

  local alarm = alerter:isalarm() or adu:isalarm()
  local alert = adu:isalertplaying()
  RailWorks.SetControlValue("AlerterAudible", 0, Misc.intbool(alarm or alert))
end

local function setcutin()
  if Misc.isinitialized() then
    local atcon = RailWorks.GetControlValue("ATCCutIn", 0) == 1
    local acseson = RailWorks.GetControlValue("ACSESCutIn", 0) == 1
    adu:setatcstate(atcon)
    adu:setacsesstate(acseson)
    alerter:setrunstate(atcon or acseson)
  end
end

local function setdynamicbraketab()
  local setuptime_s = 7
  local setuppos = 0.444444
  local insetup = RailWorks.GetControlValue("ThrottleAndBrake", 0) < setuppos
  local clock_s = RailWorks.GetSimulationTime()
  if RailWorks.GetControlValue("ThrottleAndBrake", 0) >= 0.5 then
    lastthrottletime_s = clock_s
  end
  if lastthrottletime_s ~= nil and clock_s < lastthrottletime_s + setuptime_s and
    insetup then
    RailWorks.SetControlValue("ThrottleAndBrake", 0, setuppos)
    RailWorks.SetControlValue("Buzzer", 0, 1)
  else
    RailWorks.SetControlValue("Buzzer", 0, 0)
  end
end

local function setadu()
  local aspect = adu:getaspect()
  local c, l, m, r
  if aspect == GenesisAdu.aspect.restrict then
    c, l, m, r = 0, 0, 0, 1
  elseif aspect == GenesisAdu.aspect.medium then
    c, l, m, r = 0, 0, 1, 0
  elseif aspect == GenesisAdu.aspect.limited then
    c, l, m, r = 0, 1, 0, 0
  elseif aspect == GenesisAdu.aspect.clear then
    c, l, m, r = 1, 0, 0, 0
  end
  RailWorks.SetControlValue("ADU00", 0, c)
  RailWorks.SetControlValue("ADU01", 0, l)
  RailWorks.SetControlValue("ADU02", 0, l)
  RailWorks.SetControlValue("ADU03", 0, m)
  RailWorks.SetControlValue("ADU04", 0, r)
end

local function setdisplay()
  local speed_mph = RailWorks.GetControlValue("SpeedometerMPH", 0)
  RailWorks.SetControlValue("SpeedoHundreds", 0, Misc.getdigit(speed_mph, 2))
  RailWorks.SetControlValue("SpeedoTens", 0, Misc.getdigit(speed_mph, 1))
  RailWorks.SetControlValue("SpeedoUnits", 0, Misc.getdigit(speed_mph, 0))
  RailWorks.SetControlValue("SpeedoDecimal", 0, Misc.getdigit(speed_mph, -1))

  local overspeed_mph = adu:getoverspeed_mph()
  if overspeed_mph == nil then
    RailWorks.SetControlValue("TrackHundreds", 0, -1)
    RailWorks.SetControlValue("TrackTens", 0, -1)
    RailWorks.SetControlValue("TrackUnits", 0, -1)
  else
    RailWorks.SetControlValue("TrackHundreds", 0,
                              Misc.getdigit(overspeed_mph, 2))
    RailWorks.SetControlValue("TrackTens", 0, Misc.getdigit(overspeed_mph, 1))
    RailWorks.SetControlValue("TrackUnits", 0, Misc.getdigit(overspeed_mph, 0))
  end

  RailWorks.SetControlValue("AlerterVisual", 0, Misc.intbool(alerter:isalarm()))
end

local function setplayerditchlights()
  local horntime_s = 30
  local horn = lasthorntime_s ~= nil and RailWorks.GetSimulationTime() <=
                 lasthorntime_s + horntime_s
  local flash = horn
  local headlights = RailWorks.GetControlValue("Headlights", 0)
  local crosslights = RailWorks.GetControlValue("CrossingLight", 0)
  local fixed = headlights > 0.5 and headlights < 1.5 and crosslights == 1 and
                  not flash
  ditchflasher:setflashstate(flash)
  local flashleft = ditchflasher:ison()

  local showleft = fixed or (flash and flashleft)
  RailWorks.ActivateNode("ditch_left", showleft)
  Call("DitchLight_L:Activate", Misc.intbool(showleft))

  local showright = fixed or (flash and not flashleft)
  RailWorks.ActivateNode("ditch_right", showright)
  Call("DitchLight_R:Activate", Misc.intbool(showright))
end

local function sethelperditchlights()
  RailWorks.ActivateNode("ditch_left", false)
  Call("DitchLight_L:Activate", Misc.intbool(false))
  RailWorks.ActivateNode("ditch_right", false)
  Call("DitchLight_R:Activate", Misc.intbool(false))
end

local function setaiditchlights()
  local ishead = RailWorks.GetSpeed() > Misc.stopped_mps
  RailWorks.ActivateNode("ditch_left", ishead)
  Call("DitchLight_L:Activate", Misc.intbool(ishead))
  RailWorks.ActivateNode("ditch_right", ishead)
  Call("DitchLight_R:Activate", Misc.intbool(ishead))
end

local function setcablights()
  local function activate(v) return Misc.intbool(v > 0.8) end
  -- engineer's side task light
  Call("CabLight_R:Activate",
       activate(RailWorks.GetControlValue("CabLight3", 0)))
  -- engineer's forward task light
  Call("TaskLight_R:Activate",
       activate(RailWorks.GetControlValue("CabLight1", 0)))
  -- secondman's forward task light
  Call("TaskLight_L:Activate",
       activate(RailWorks.GetControlValue("CabLight2", 0)))
  -- secondman's side task light
  Call("CabLight_L:Activate",
       activate(RailWorks.GetControlValue("CabLight4", 0)))
  -- dome light
  Call("CabLight_M:Activate",
       activate(RailWorks.GetControlValue("CabLight5", 0)))
end

local function setexhaust()
  local r, g, b, rate
  local minrpm = 180
  local effort = RailWorks.GetTractiveEffort()
  if RailWorks.GetControlValue("RPM", 0) < minrpm then
    r, g, b = 0, 0, 0
    rate = 0
    -- DTG's exhaust logic
  elseif effort < 0.1 then
    r, g, b = 0.25, 0.25, 0.25
    rate = 0.01
  elseif effort >= 0.1 and effort < 0.5 then
    r, g, b = 0.1, 0.1, 0.1
    rate = 0.005
  else
    r, g, b = 0, 0, 0
    rate = 0.001
  end
  Call("DieselExhaust:SetEmitterColour", r, g, b)
  Call("DieselExhaust:SetEmitterRate", rate)
end

local function updateplayer(dt)
  readcontrols()

  adu:update(dt)
  alerter:update(dt)
  blight:playerupdate(dt)

  writelocostate()
  setcutin()
  setdynamicbraketab()
  setadu()
  setdisplay()
  setplayerditchlights()
  setcablights()
  setexhaust()
end

local function updatehelper()
  sethelperditchlights()
  setcablights()
  setexhaust()
end

local function updateai()
  setaiditchlights()
  setcablights()
  setexhaust()
end

Update = Misc.wraperrors(function(dt)
  if RailWorks.GetIsEngineWithKey() then
    updateplayer(dt)
  elseif RailWorks.GetIsPlayer() then
    updatehelper(dt)
  else
    updateai()
  end
end)

OnControlValueChange = Misc.wraperrors(function(name, index, value)
  if name == "ThrottleAndBrake" or name == "TrainBrakeControl" then
    alerter:acknowledge()
  end

  RailWorks.SetControlValue(name, index, value)
end)

OnCustomSignalMessage = Misc.wraperrors(function(message)
  adu:receivemessage(message)
end)

OnConsistMessage = Misc.wraperrors(function(message, argument, direction)
  blight:receivemessage(message, argument, direction)

  RailWorks.Engine_SendConsistMessage(message, argument, direction)
end)

File: src/mod/Assets/DTG/WashingtonBaltimore/RailVehicles/Diesel/P42DC/Amtrak/Locomotive/P42DC_EngineScript.ts
/**
 * Amtrak GE P42DC
 */

import * as ale from "lib/alerter";
import * as c from "lib/constants";
import * as frp from "lib/frp";
import { FrpEngine } from "lib/frp-engine";
import { mapBehavior, rejectRepeats } from "lib/frp-extra";
import { SensedDirection } from "lib/frp-vehicle";
import * as m from "lib/math";
import { AduAspect } from "lib/nec/adu";
import * as cs from "lib/nec/cabsignals";
import * as adu from "lib/nec/twospeed-adu";
import * as rw from "lib/railworks";
import * as fx from "lib/special-fx";
import * as ui from "lib/ui";

const ditchLightsFadeS = 0.3;
const ditchLightFlashS = 0.65;
const taskLightFadeS = 0.8 / 2;

const me = new FrpEngine(() => {
    // Safety systems cut in/out
    const atcCutIn = () => (me.rv.GetControlValue("Window Left") as number) < 0.9;
    const acsesCutIn = () => (me.rv.GetControlValue("Window Right") as number) < 0.9;
    ui.createAtcStatusPopup(me, atcCutIn);
    ui.createAcsesStatusPopup(me, acsesCutIn);
    const alerterCutIn = frp.liftN((atcCutIn, acsesCutIn) => atcCutIn || acsesCutIn, atcCutIn, acsesCutIn);
    ui.createAlerterStatusPopup(me, alerterCutIn);

    // Safety systems and ADU
    const acknowledge = me.createAcknowledgeBehavior();
    const suppression = () => (me.rv.GetControlValue("TrainBrakeControl") as number) >= 0.4;
    const [aduState$, aduEvents$] = adu.create({
        atc: cs.fourAspectAtc,
        e: me,
        acknowledge,
        suppression,
        atcCutIn,
        acsesCutIn,
        equipmentSpeedMps: 110 * c.mph.toMps,
    });
    const aduStateHub$ = frp.compose(aduState$, frp.hub());
    aduStateHub$(({ aspect, trackSpeedMph }) => {
        me.rv.SetControlValue("ADU00", aspect === cs.FourAspect.Clear ? 1 : 0);
        me.rv.SetControlValue("ADU01", aspect === cs.FourAspect.ApproachLimited ? 1 : 0);
        me.rv.SetControlValue("ADU02", aspect === cs.FourAspect.ApproachLimited ? 1 : 0);
        me.rv.SetControlValue("ADU03", aspect === cs.FourAspect.Approach ? 1 : 0);
        me.rv.SetControlValue("ADU04", aspect === cs.FourAspect.Restricting || aspect === AduAspect.Stop ? 1 : 0);

        if (trackSpeedMph !== undefined) {
            const [[h, t, u]] = m.digits(trackSpeedMph, 3);
            me.rv.SetControlValue("TrackHundreds", h);
            me.rv.SetControlValue("TrackTens", t);
            me.rv.SetControlValue("TrackUnits", u);
        } else {
            me.rv.SetControlValue("TrackHundreds", -1);
            me.rv.SetControlValue("TrackTens", -1);
            me.rv.SetControlValue("TrackUnits", -1);
        }
    });
    const aduState = frp.stepper(aduStateHub$, undefined);
    // Alerter
    const alerter$ = frp.compose(ale.create({ e: me, acknowledge, cutIn: alerterCutIn }), frp.hub());
    const alerterState = frp.stepper(alerter$, undefined);
    alerter$(({ alarm }) => {
        me.rv.SetControlValue("AlerterVisual", alarm ? 1 : 0);
    });
    // Safety system sounds
    const upgradeEvents$ = frp.compose(
        aduEvents$,
        frp.filter(evt => evt === adu.AduEvent.Upgrade)
    );
    const upgradeSound$ = frp.compose(me.createPlayerWithKeyUpdateStream(), fx.triggerSound(0.25, upgradeEvents$));
    const isAlarm$ = frp.compose(
        me.createPlayerWithKeyUpdateStream(),
        mapBehavior(
            frp.liftN(
                (aduState, alerterState, upgradeSound) =>
                    aduState?.atcAlarm || aduState?.acsesAlarm || alerterState?.alarm || upgradeSound,
                aduState,
                alerterState,
                frp.stepper(upgradeSound$, false)
            )
        )
    );
    isAlarm$(on => {
        me.rv.SetControlValue("AlerterAudible", on ? 1 : 0);
    });

    // Throttle, dynamic brake, and air brake controls
    const isPenaltyBrake = frp.liftN(
        (aduState, alerterState) => (aduState?.penaltyBrake || alerterState?.penaltyBrake) ?? false,
        aduState,
        alerterState
    );
    // There's no virtual throttle, so just move the power handle.
    const setCombinedPower$ = frp.compose(
        me.createPlayerWithKeyUpdateStream(),
        frp.filter(_ => frp.snapshot(isPenaltyBrake)),
        frp.map(_ => 0.5)
    );
    setCombinedPower$(v => {
        me.rv.SetControlValue("ThrottleAndBrake", v);
    });
    // There's no virtual train brake, so just move the braking handle.
    const fullService = 0.85;
    const setPenaltyBrake$ = frp.compose(
        me.createPlayerWithKeyUpdateStream(),
        frp.filter(_ => frp.snapshot(isPenaltyBrake)),
        frp.map(_ => fullService)
    );
    setPenaltyBrake$(v => {
        me.rv.SetControlValue("TrainBrakeControl", v);
    });
    // Dynamic brake lockout
    const combinedPower = () => me.rv.GetControlValue("ThrottleAndBrake") as number;
    const setupNotch = 0.444444;
    const dynamicLockout$ = frp.compose(
        me.createPlayerWithKeyUpdateStream(),
        fx.behaviorStopwatchS(frp.liftN(cp => cp < 0.5, combinedPower)),
        frp.map(brakingS => frp.snapshot(combinedPower) < setupNotch && brakingS !== undefined && brakingS < 7)
    );
    dynamicLockout$(lock => {
        if (lock) {
            me.rv.SetControlValue("ThrottleAndBrake", setupNotch);
        }
        me.rv.SetControlValue("Buzzer", lock ? 1 : 0);
    });

    // Ditch lights
    const ditchLights = [
        new fx.FadeableLight(me, ditchLightsFadeS, "DitchLight_L"),
        new fx.FadeableLight(me, ditchLightsFadeS, "DitchLight_R"),
    ];
    const areHeadLightsOn = () => {
        const cv = me.rv.GetControlValue("Headlights") as number;
        return cv > 0.5 && cv < 1.5;
    };
    const areDitchLightsOn = frp.liftN(
        (headLights, crossingLight) => headLights && crossingLight,
        areHeadLightsOn,
        () => (me.rv.GetControlValue("CrossingLight") as number) > 0.5
    );
    const hornSequencer$ = frp.compose(me.createPlayerWithKeyUpdateStream(), me.mapGetCvStream("CylinderCock"));
    const ditchLightsHornTimeS$ = frp.compose(
        me.createPlayerWithKeyUpdateStream(),
        me.mapGetCvStream("Horn"),
        frp.merge(hornSequencer$),
        frp.filter(v => v === 1),
        frp.map(_ => me.e.GetSimulationTime())
    );
    const ditchLightsFlash = frp.liftN(
        (ditchOn, bell, hornTimeS) =>
            ditchOn && (bell || (hornTimeS !== undefined && me.e.GetSimulationTime() - hornTimeS <= 30)),
        areDitchLightsOn,
        () => (me.rv.GetControlValue("Bell") as number) > 0.5,
        frp.stepper(ditchLightsHornTimeS$, undefined)
    );
    const ditchLightsPlayer$ = frp.compose(
        me.createPlayerWithKeyUpdateStream(),
        fx.behaviorStopwatchS(ditchLightsFlash),
        frp.map((stopwatchS): [boolean, boolean] => {
            if (stopwatchS === undefined) {
                const ditchOn = frp.snapshot(areDitchLightsOn);
                return [ditchOn, ditchOn];
            } else {
                const flashLeft = stopwatchS % (ditchLightFlashS * 2) < ditchLightFlashS;
                return [flashLeft, !flashLeft];
            }
        })
    );
    const ditchLightsAi$ = frp.compose(
        me.createAiUpdateStream(),
        frp.map((au): [boolean, boolean] => {
            const ditchOn = au.direction === SensedDirection.Forward;
            return [ditchOn, ditchOn];
        })
    );
    const ditchLights$ = frp.compose(
        me.createPlayerWithoutKeyUpdateStream(),
        frp.map((_): [boolean, boolean] => [false, false]),
        frp.merge(ditchLightsPlayer$),
        frp.merge(ditchLightsAi$)
    );
    ditchLights$(([l, r]) => {
        const [left, right] = ditchLights;
        left.setOnOff(l);
        right.setOnOff(r);
    });
    const ditchNodeLeft$ = frp.compose(
        me.createUpdateStream(),
        frp.map(_ => {
            const [light] = ditchLights;
            return light.getIntensity() > 0.5;
        }),
        rejectRepeats()
    );
    const ditchNodeRight$ = frp.compose(
        me.createUpdateStream(),
        frp.map(_ => {
            const [, light] = ditchLights;
            return light.getIntensity() > 0.5;
        }),
        rejectRepeats()
    );
    ditchNodeLeft$(on => {
        me.rv.ActivateNode("ditch_left", on);
    });
    ditchNodeRight$(on => {
        me.rv.ActivateNode("ditch_right", on);
    });

    // Horn rings the bell.
    const bellControl$ = frp.compose(
        me.createOnCvChangeStreamFor("Horn"),
        frp.merge(me.createOnCvChangeStreamFor("CylinderCock")),
        frp.filter(v => v === 1),
        me.mapAutoBellStream()
    );
    bellControl$(v => {
        me.rv.SetControlValue("Bell", v);
    });

    // Diesel exhaust effect
    const exhaust = new rw.Emitter("DieselExhaust");
    const exhaustState = frp.liftN(
        (rpm, effort): [color: number, rate: number] => {
            // DTG's exhaust logic
            if (rpm < 180) {
                return [0, 0];
            } else if (effort < 0.1) {
                return [0.25, 0.01];
            } else if (effort < 0.5) {
                return [0.1, 0.005];
            } else {
                return [0, 0.001];
            }
        },
        () => me.rv.GetControlValue("RPM") as number,
        () => me.eng.GetTractiveEffort()
    );
    const exhaustActive$ = frp.compose(
        me.createUpdateStream(),
        mapBehavior(exhaustState),
        frp.map(([, rate]) => rate > 0),
        rejectRepeats()
    );
    const exhaustColor$ = frp.compose(
        me.createUpdateStream(),
        mapBehavior(exhaustState),
        frp.map(([color]) => color),
        rejectRepeats()
    );
    const exhaustRate$ = frp.compose(
        me.createUpdateStream(),
        mapBehavior(exhaustState),
        frp.map(([, rate]) => rate),
        rejectRepeats()
    );
    exhaustActive$(active => {
        exhaust.SetEmitterActive(active);
    });
    exhaustColor$(color => {
        exhaust.SetEmitterColour(color, color, color);
    });
    exhaustRate$(rate => {
        exhaust.SetEmitterRate(rate);
    });

    // Speedometer
    const aSpeedoMph$ = frp.compose(
        me.createPlayerWithKeyUpdateStream(),
        mapBehavior(me.createSpeedometerMpsBehavior()),
        frp.map(mps => mps * c.mps.toMph),
        frp.map(mph => Math.abs(mph))
    );
    aSpeedoMph$(mph => {
        const [[h, t, u]] = m.digits(Math.floor(mph), 3);
        me.rv.SetControlValue("SpeedoHundreds", h);
        me.rv.SetControlValue("SpeedoTens", t);
        me.rv.SetControlValue("SpeedoUnits", u);
        me.rv.SetControlValue("SpeedoDecimal", m.tenths(mph));
    });

    // Cockpit lights
    // Engineer's side task light
    createTaskLight(me, "CabLight_R", "CabLight3");
    // Engineer's forward task light
    createTaskLight(me, "TaskLight_R", "CabLight1");
    // Secondman's forward task light
    createTaskLight(me, "TaskLight_L", "CabLight2");
    // Secondman's side task light
    createTaskLight(me, "CabLight_L", "CabLight4");
    // Dome light
    createTaskLight(me, "CabLight_M", "CabLight5");

    // Process OnControlValueChange events.
    const onCvChange$ = frp.compose(
        me.createOnCvChangeStream(),
        frp.reject(([name]) => name === "Bell")
    );
    onCvChange$(([name, value]) => {
        me.rv.SetControlValue(name, value);
    });

    // Set consist brake lights.
    fx.createBrakeLightStreamForEngine(me);

    // Set platform door height.
    fx.createLowPlatformStreamForEngine(me, true);

    // Enable updates.
    me.e.BeginUpdate();

    // Set engine number.
    const [[lh, lt, lu]] = m.digits(tonumber(me.rv.GetRVNumber()) ?? 0, 3);
    me.rv.SetControlValue("LocoHundreds", lh);
    me.rv.SetControlValue("LocoTens", lt);
    me.rv.SetControlValue("LocoUnits", lu);
});
me.setup();

function createTaskLight(e: FrpEngine, light: string, cv: string) {
    const fadeableLight = new fx.FadeableLight(e, taskLightFadeS, light);
    const stream$ = frp.compose(
        e.createPlayerWithKeyUpdateStream(),
        e.mapGetCvStream(cv),
        frp.map(v => v > 0.5)
    );
    stream$(on => {
        fadeableLight.setOnOff(on);
    });
}
